Integrate KeyboardioScanner as a driver

algernon opened the pull request:
In the process of moving towards a single repository for everything Kaleidoscope, integrate KeyboardioScanner as a driver. This is a direct copy of KeyboardioScanner as of keyboardio/KeyboardioScanner@2090cd426cae25b07c0ce3a6b7365b95c21dd87b, renamed and namespaced to fit into Kaleidoscope.

I opted to not take the full history of KeyboardioScanner, but copy it instead.

There are some differences in PROGMEM & RAM size: compared to master, we're at -42 PROGMEM, at the cost of +160 RAM. I'm not entirely sure yet what the extra RAM is for, but while a lot, it doesn't seem to be _too_ much.

algernon commented:
Oh, I know! That 160 RAM is `twi_masterBuffer`, which defaults to 192 bytes in `kaleidoscope/device/keyboardio/twi.c`, but to 32 in KeyboardioScanner's twi.c, and I opted to not have another copy of twi.c.

So it boils down to making it possible to set the buffer size in a saner way, which likely involves c++-izing twi.c, and making it templateable. I'd do that in a followup PR.

algernon commented:
We also have a filename conflict, between `device::dygma::raise::Hand` and `driver::keyscanner::keyboardio::model01::Hand`, and that's a valid concern, because the driver does not have hardware-specific ifdef guards, so if compiling for the Raise, we would, indeed, have a conflict - but not anywhere else, because `raise::Hand` _is_ guarded by a hardware-specific ifdef.

@obra Do you have a suggestion for naming, by any chance?

algernon commented:
> That 160 RAM is twi_masterBuffer, which defaults to 192 bytes in kaleidoscope/device/keyboardio/twi.c, but to 32 in KeyboardioScanner's twi.c, and I opted to not have another copy of twi.c.

Another option is to just cut it down to 32, since the only other user of twi.c at the moment is the Keyboardio Imago, and I think the Model01 is more important at this point. Possibly breaking Imago for some time sounds acceptable to me.

obra commented:
> On Dec 14, 2020, at 12:01 PM, Gergely Nagy <notifications@github.com> wrote:
> 
> 
> That 160 RAM is twi_masterBuffer, which defaults to 192 bytes in kaleidoscope/device/keyboardio/twi.c, but to 32 in KeyboardioScanner's twi.c, and I opted to not have another copy of twi.c.
> 
> Another option is to just cut it down to 32, since the only other user of twi.c at the moment is the Keyboardio Imago, and I think the Model01 is more important at this point. Possibly breaking Imago for some time sounds acceptable to me.frm 
> 

That’s just kind of kicking the problem down the road. Instead, how about allowing a default value to be overridden by a -D value on the board.


> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub <https://github.com/keyboardio/Kaleidoscope/pull/994#issuecomment-744676676>, or unsubscribe <https://github.com/notifications/unsubscribe-auth/AAALC2CSJHE33AZ727YILGDSUZVK7ANCNFSM4U3DSI4A>.
>

obra commented:
> On Dec 14, 2020, at 10:08 AM, Gergely Nagy <notifications@github.com> wrote:
> 
> 
> We also have a filename conflict, between device::dygma::raise::Hand and driver::keyscanner::keyboardio::model01::Hand, and that's a valid concern, because the driver does not have hardware-specific ifdef guards, so if compiling for the Raise, we would, indeed, have a conflict - but not anywhere else, because raise::Hand is guarded by a hardware-specific ifdef.
> 
> @obra <https://github.com/obra> Do you have a suggestion for naming, by any chance?
> 
> 

At least as of now, the exact same driver will be used for the Model 100.

What do weed to do to be able to have multiple source files with the same name in different directories or namespaces?
> —

algernon commented:
> On Dec 14, 2020, at 12:01 PM, Gergely Nagy ***@***.***> wrote: That 160 RAM is twi_masterBuffer, which defaults to 192 bytes in kaleidoscope/device/keyboardio/twi.c, but to 32 in KeyboardioScanner's twi.c, and I opted to not have another copy of twi.c. Another option is to just cut it down to 32, since the only other user of twi.c at the moment is the Keyboardio Imago, and I think the Model01 is more important at this point. Possibly breaking Imago for some time sounds acceptable to me.frm
> That’s just kind of kicking the problem down the road. Instead, how about allowing a default value to be overridden by a -D value on the board.

That's already the case, the default is guarded by an `#ifndef`. So yeah, we can just add a -D to the board, too.

obra commented:
Side item: I know we have a gamma driver, but see also that we still have the gamma table for this driver in here. Is that fixable?

algernon commented:
> At least as of now, the exact same driver will be used for the Model 100.

*nod*. Any suggestion for the namespace? `driver::keyscanner::keyboardio::model101`, for both model100 and model01?

> What do weed to do to be able to have multiple source files with the same name in different directories or namespaces?

Basically, wait for a miracle. :(

The problem is that we use static archives (`dot_a_linkage=true` in `library.properties), because that saves us a ton of progmem. However, the static archives are super simple, each object is stored as its basename, without any paths. So if you have `a/foo.o` and `b/foo.o`, both of them will be stored as `foo.o` in the archive, and whichever gets linked later, will override the former. This has caused us problems in the past (hence the CI check for it). The linker can't be augmented to store paths, so unless we drastically change our file structure, by encoding the path in filenames too, so `driver/keyscanner/keyboardio/model01/Hand.cpp` becomes `driver/keyscanner/keyboardio/model01/driver_keyscanner_keyboardio_model01_Hand.cpp` (either doing this manually, or automated during build or something), we will have conflicts.

I have a couple of ideas we could explore, to make this problem less of an issue, but... lets talk about that outside of this ticket.

In the short-term, I'll need to rename one of the `Hand.[cpp|h]` files. I suppose `Side.[cpp|h]` would work?

> Side item: I know we have a gamma driver, but see also that we still have the gamma table for this driver in here. Is that fixable?

Oh, yeah. It is. Will do that in a followup commit.

algernon commented:
> Any suggestion for the namespace? driver::keyscanner::keyboardio::model101, for both model100 and model01?

Perhaps just `driver::keyscanner::keyboardio::<something>`? (where `<something>` is whatever we end up renaming `Hand.cpp` to)

obra commented:
I think that we'd be better off changing both of them, to something like Model01Side and RaiseSide.

obra commented:
(Also, it’s not really a keyscanner driver - one of the things it offers is keyscanning, but since it’s also an led driver….it may not belong down in that namespace)

> On Dec 14, 2020, at 12:23 PM, Gergely Nagy <notifications@github.com> wrote:
> 
> 
> Any suggestion for the namespace? driver::keyscanner::keyboardio::model101, for both model100 and model01?
> 
> Perhaps just driver::keyscanner::keyboardio::<something>? (where <something> is whatever we end up renaming Hand.cpp to)
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub <https://github.com/keyboardio/Kaleidoscope/pull/994#issuecomment-744688417>, or unsubscribe <https://github.com/notifications/unsubscribe-auth/AAALC2ADDWESF4JBZKVTHMDSUZX5DANCNFSM4U3DSI4A>.
>

algernon commented:
`*Side` it is then. As for the namespace, yeah... I guess I can do `driver::keyboardio::` and just put it there.

Thanks!

algernon commented:
- Moved the driver to `kaleidoscope::driver::keyboardio::Model01Side`
- Removed the copy of the gamma8 table, we're now using the gamma correction table already in Kaleidoscope
- raise::Hand got renamed to raise::RaiseSide
